src/server/server_family.cc

```diff
@@ -1930,15 +1930,34 @@ void ServerFamily::ReplConf(CmdArgList args, ConnectionContext* cntx) {
       auto& pool = service_.proactor_pool();
       pool.AwaitFiberOnAll(
           [&](util::ProactorBase* pb) { ServerState::tlocal()->remote_client_id_ = arg; });
+    } else if (cmd == "ACK" && args.size() == 2) {
+      // Don't send error/Ok back through the socket, because we don't want to interleave with
+      // the journal writes that we write into the same socket.
+
+      if (!cntx->replication_flow) {
+        LOG(ERROR) << "No replication flow assigned";
+        return;
+      }
+
+      uint64_t ack;
+      if (!absl::SimpleAtoi(arg, &ack)) {
+        LOG(ERROR) << "Bad int in REPLCONF ACK command! arg=" << arg;
+        return;
+      }
+      VLOG(1) << "Received client ACK=" << ack;
+      cntx->replication_flow->last_acked_lsn = ack;
+      return;
```

Comment:
_I am not sure at all_, but I was thinking if there is any option for invalid pointer access.

If a flow fails, master starts cancellation. Part of cancellation on stable sync is running flow cleanup, it closes the socket. Then replica_ptr is dropped, so all flows are deallocated. Closing the socket will make the connection close.

Is there a way for the ack to be received before the socket is closed, but run after the flow has been deleted (and the context isn't closed yet)? Some unlucky preemption inside connection? It seems like there are none like this. But it would be a wild bug

Reply:
I think it's Ok:
We call `flow->cleanup()` and `flow->full_sync_fb.Join()` in `DflyCmd::CancelReplication` before erasing the `ReplicationInfo` struct from `DflyCmd::replica_infos_`